neurobooth_os/iout/mbient.py

```diff
@@ -110,58 +295,196 @@ def reset_device(device: MetaWear) -> None:
+        self.outlet_id = str(uuid.uuid4())
+        self.buzz_time = buzz_time_sec
+        self.max_connect_attempts = try_nmax
+        self.retry_delay_sec = 1
+
+        # Device configuration settings
+        self.connection_params = ConnectionParameters()  # Use the default params
+        self.accel_params = SensorParameters(sample_rate=acc_hz, data_range=16.0)
+        self.gyro_params = SensorParameters(sample_rate=gyro_hz, data_range=2000)
+
+        # Uninitialized Variables
+        self.device: Optional[MetaWear] = None
+        self.subscribed_signals: List[Any] = []
+        self.outlet: Optional[StreamOutlet] = None
+        self.callback: Callable = lambda *args: None
+
+        # Streaming-related variables
+        self.streaming: bool = False
+        self.n_samples_streamed = 0
 
         self.logger = logging.getLogger('session')
-        self.logger.debug(f'Mbient [{self.dev_name}]: acc_sample_rate={self.acc_hz}; gyro_sample_rate={self.gyro_hz}')
+        self.logger.debug(self.format_message(f'acc={self.accel_params}; gyro={self.gyro_params}'))
+
+    def format_message(self, msg: str) -> str:
+        return f'Mbient [{self.dev_name}; {self.mac}]: {msg}'
+
+    def prepare_scan(self) -> None:
+        """
+        Perform a BLE scan to wake up devices before trying to connect.
+        (The alternative is to physically push the button on the devices.)
+        We only need to do this once, so this function ensures it is only done once per machine/server.
+        """
+        with self.SCAN_LOCK:
+            if self.SCAN_PERFORMED:  # Only need to scan once if multiple devices are present
+                return
+            self.logger.debug('Performing BLE Scan')
+            ble_devices = scan_BLE(timeout_sec=10)
+            self.logger.debug(f'BLE scan found {len(ble_devices)} devices: {[mac for _, mac in ble_devices.items()]}')
+            self.SCAN_PERFORMED = True
+
+    def connect(self, n_attempts: Optional[int] = None, retry_delay_sec: Optional[float] = None) -> None:
+        """
+        Attempt to connect to the device and set a disconnect handler.
+
+        :param n_attempts: How many times to attempt a connection before giving up.
+        :param retry_delay_sec: How long to wait in-between attempts.
+        """
+        if n_attempts is None:
+            n_attempts = self.max_connect_attempts
+        if retry_delay_sec is None:
+            retry_delay_sec = self.retry_delay_sec
+
+        self.device = connect_device(
+            mac_address=self.mac,
+            n_attempts=n_attempts,
+            retry_delay_sec=retry_delay_sec,
+            log_fn=lambda msg: self.logger.debug(self.format_message(msg)),
+        )
+        self.device.on_disconnect = lambda status: self.on_disconnect(status)
+
+    def on_disconnect(self, status=None) -> None:
+        """
+        Callback for disconnect events. Attempt to reconnect to and configure the device.
```

Comment:
Is this only for premature disconnects (as suggested by the print statement below)? If so, maybe rename?

Reply:
Done